onnxruntime/core/session/custom_ops.cc

```diff
@@ -419,120 +677,48 @@ common::Status CreateCustomRegistry(gsl::span<OrtCustomOpDomain* const> op_domai
-        // ORT API version 8. So, query the relevant methods ONLY from API version 8 onwards.
-        if (op->version >= min_ort_version_with_optional_io_support) {
-          const auto characteristic = op->GetOutputCharacteristic(op, i);
-
-          // Support for optional and variadic inputs/output was added in versions 8 and 14, respectively.
-          if (characteristic == OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_OPTIONAL) {
-            option = onnx::OpSchema::FormalParameterOption::Optional;
-          } else if ((op->version >= min_ort_version_with_variadic_io_support) &&
-                     (characteristic == OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_VARIADIC)) {
-            ORT_ENFORCE(i == output_count - 1, "Only the last output to a custom op may be marked variadic.");
-            option = onnx::OpSchema::FormalParameterOption::Variadic;
-            min_arity = op->GetVariadicOutputMinArity(op);
-            is_homogeneous = static_cast<bool>(op->GetVariadicOutputHomogeneity(op));
-          }
-        }
-
-        const auto type = op->GetOutputType(op, i);
-        if (ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED == type) {  // Dynamic typed output
-          if (op->GetOutputCharacteristic(op, i) == OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_REQUIRED) {
-              ORT_ENFORCE(type_id_counter == 1,
-                          "There must be one (and only one) dynamic typed input to the custom op. "
-                          "Its type info at runtime will be used to infer the type info of this dynamic typed output "
-                          "which is required for the success of the model loading step. "
-                          "More than one dynamic typed inputs are currently not supported as differing types at runtime means the output type "
-                          "cannot be inferred without which model loading cannot proceed.");
-          }
-
-          schema.Output(i, "Output" + std::to_string(i), "", "T0", option, is_homogeneous, min_arity);
-        } else {
-          schema.Output(i, "Output" + std::to_string(i), "",
-                        DataTypeImpl::ToString(onnxruntime::DataTypeImpl::TensorTypeFromONNXEnum(type)), option,
-                        is_homogeneous, min_arity);
-        }
+      // define kernel
+      auto kernel_create_info = CreateKernelCreateInfo(domain->domain_, op);
+      kernel_def_map[op->GetName(op)].push_back(kernel_create_info.kernel_def.get());
+      ORT_RETURN_IF_ERROR(output->RegisterCustomKernel(kernel_create_info));
+      // define schema
+      auto schema_map_iter = schema_map.find(op->GetName(op));
+      if (schema_map_iter == schema_map.end()) {
+        auto schema = CreateSchema(domain->domain_, op);
+        schema_map.emplace(schema.Name(), schema);
+      } else {
+        ORT_RETURN_IF_ERROR(IsCompatible(schema_map_iter->second, op));
```

Comment:
should we add a check to prevent user register two custom kernel with the same type?

Reply:
Added a corresponding unit test there for the case.
Seems kernel-registry does not mind duplicated customer kernels.
For schema, by this PR there got to be only one that is getting created.